src/wallet/rpcwallet.cpp

```diff
@@ -3374,55 +3391,128 @@ UniValue z_listreceivedbyaddress(const UniValue& params, bool fHelp)
     }
 
     UniValue result(UniValue::VARR);
-    std::vector<SproutNoteEntry> sproutEntries;
-    std::vector<SaplingNoteEntry> saplingEntries;
-    pwalletMain->GetFilteredNotes(sproutEntries, saplingEntries, fromaddress, nMinDepth, false, false);
 
     std::set<std::pair<PaymentAddress, uint256>> nullifierSet;
     auto hasSpendingKey = boost::apply_visitor(HaveSpendingKeyForPaymentAddress(pwalletMain), zaddr);
     if (hasSpendingKey) {
         nullifierSet = pwalletMain->GetNullifiersForAddresses({zaddr});
     }
 
+    NotesFilters filters;
+    filters.minDepth = nMinDepth;
+    filters.ignoreSpent = false;
+    filters.requireSpendingKey = false;
+    filters.cwallet = pwalletMain;
+
     if (boost::get<libzcash::SproutPaymentAddress>(&zaddr) != nullptr) {
-        for (SproutNoteEntry & entry : sproutEntries) {
+        const auto pa = *boost::get<libzcash::SproutPaymentAddress>(&zaddr);
+        const auto itr_sprout = pwalletMain->GetNotesByType(pa.GetHash(), NoteType::sprout, timestamp);
+        auto itr = itr_sprout.begin();
+
+        if (offset > 0) {
+            if(offset < (int)std::distance(itr, itr_sprout.end())) {
+                boost::advance(itr, offset);
```

Comment:
I don't think you should use `boost::advance()` here. 
Imagine you are supposed to skip 10 records, but these 10 records include 3 notes that would have been skipped by the fitlers.

So used is iterating over the whole dataset in chunks of 10. The whole dataset has 15 notes, but 3 are excluded by filters and 12 should be returned to the client.

First chunk would yield: [n1, n2, n4, n6, n7, n8, n10, n11, n12, 13], because n3, n6 and n5 where excluded by filters.
Second chunk would yield: [n11, n12, n13, n14, n15]. It appears as if 15 notes where returned, but n11, n12, n13 where returned as duplicates.

Reply:
Yea, i see what you mean. I will prefer to just don't do it inside the loop but will do it if no other alternative. 

What if we add an `id` field to the index making it `type, pa, timestamp, id`. Offset will be replaced by id in rpc calls, `id` is just an auto increment added to the helper object when they are inserted to the wallet as they came.

In your example, each node will have an id so the full set will be:
`[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15]`

First call is done with id=0, client will get:
`[1, 2, 4, 7, 8, 9, 10, 11, 12, 13] `

Client must use the last id obtained for pagination, in this case `13` so client next call will be at 13 + 1 = 14,  response should be:

`[14, 15]`

Let me know how do you see it.

Reply:
I think it's good direction, but details.

The list get's returned sorted by timestamp. There is another comment I put in the first review, where I didn't exactly understand *which* timestamp are you using, because you just seemed to do `GetTime()` at the moment of inserting into the index, which may mean different things.

Putting this aside, I agree with you that perhaps we don't need `offset` at all, but I wouldn't introduce `id` either.
Instead I would just use `timestamp`. When client wants to iterate over the whole dataset in chunks he would just do the query with sinceTimestamp=0, limit=10.
Lets say he would get a list of 10 result. To get next page he would take timestamp of last entry and do the query with sinceTimestamp=result[limit - 1].timestamp, limit=10 and continue doing requests until the obtained length(result) < limit.
I think this would work just fine.

However there are few conditions: 
* timestamp have to be unique - if there are 2 entries with exactly same timestamp and they happen on the boundary between pages, we would skip the second entry.
* sinceTimestamp need to work with `>` and not with `>=` because otherwise we would get the last entry from previous page repeat as first entry on the next.

Please let me know what you think.

Reply:
I agree that timestamp is similar but we cant guarantee that it is unique independently if it is get time, block time or any other. 2 notes can land at the same time or at least i don't see why this could not happen. An `id` will not have this limitation.

Reply:
Well, that's why I'm asking how timestamp is determined.
For example if timestamp is taken locally using OS call, I think it's one of the guaranties that you'll never get the same return value. But if it's generated by the sender, than multiple senders may get the same results. Also if the timestamp is rounded up to seconds, than, of course there will be plenty of conflicts.

So, I'd say it depends.

In any case, I think that this ID proposal may be better, but perhaps call it something like `rownumber`, because `id` suggest that the message will have the same `id` while it's clearly not the case.

Reply:
The time used is local time as `time()` is used in `GetTime()`: https://github.com/zcash/zcash/blob/master/src/utiltime.cpp#L24

I used this as when i try to get transaction time at insertion i get 0. I didn't researched much further but i think transaction time is added when confirmation is 1 so we have no way to know blocktime when we first see a transaction at least if we generated it.

As we use it just for sorting, using local time as we insert should be ok.

Now, in the same file we have `GetTimeMillis()`: https://github.com/zcash/zcash/blob/master/src/utiltime.cpp#L32 and we can even add `GetTimeNano()` if we want.

Will this be enough to guarantee the uniqueness of the timestamp ? If so, we dont need the additional field and we can change the index to `ordered_unique`.

Reply:
I would expect that the `GetTimeNano()` should be enough and it returns unique values.
Can you write a simple program that prints the result of this function two times one by one ?

I'd do it myself, but figuring out how to link to zcash libraries is way above my cpp skills....

Reply:
Tried it, the number is always different calling one just after the other. Implemented at https://github.com/zcash/zcash/pull/4399/commits/a56fc3344d448cb3779378b1c43f7b095793eeed

Reply:
AFAIK this isn't a guaranteed property of the implementation of `GetTimeNanos()` in this PR. It might also be platform-dependent, because some platforms might only have a lower-resolution clock. (`std::chrono` doesn't make any promises about that.)

Reply:
Any suggestion for this one ?